site/docs/skylark/tutorial-custom-verbs.md

```diff
@@ -0,0 +1,157 @@
+existing rule for generating a website with Sphinx:
+
+```python
+_sphinx_site = rule(
+     implementation = _sphinx_impl,
+     attrs = {"srcs": attr.label_list(allow_files = [".rst"])},
+)
+```
+
+Next, consider a rule like the following, which builds a script that, when run,
+publishes the generated pages:
+
+```python
+_sphinx_publisher = rule(
+    implementation = _publish_impl,
+    attrs = {
+        "site": attr.label(),
+        "_publisher": attr.label(
+            default = "//internal/sphinx:publisher",
+            executable = True,
+        ),
+    },
+    executable = True,
+)
+```
+
+Finally, we define the following macro to create targets for both of the above
+rules together:
+
+```python
+def sphinx_site(name, srcs = [], **kwargs):
+    # This creates the primary target, propducing the Sphinx-generated HTML.
+    _sphinx_site(name = name, srcs = srcs, **kwargs)
+    # This creates the secondary target, which produces a script for publishing
+    # the site generated above.
+    _sphinx_publisher(name = "%s.publish" % name, site = name, **kwargs)
+```
+
+In our `BUILD` files, we use the macro as though it just creates the primary target:
+
+```python
+sphinx_site(
+    name = "docs",
+    srcs = ["index.md", "providers.md"],
+)
+```
+
+In this example, a "docs" target is created, just as though the macro were a
+standard, single Bazel rule. When built, the rule generates some configuration
+and runs Sphinx to produce an HTML site, ready for manual inspection. However,
+an additional "docs.publish" target is also created, which builds a script for
+publishing the site. Once we check the output of the primary target, we can use
+`bazel run :docs.publish` to publish it for public consumption, just like an
+imaginary `bazel publish` command.
+
+
+It's not immediately obvious what the implementation of the `_sphinx_publisher`
+rule might look like. Often, actions like this write a _launcher_ shell script.
+This method typically involves using
+[`ctx.actions.expand_template`](https://docs.bazel.build/versions/0.28.0/skylark/lib/actions.html#expand_template)
+to write a very simple shell script, in this case invoking the publisher binary
+with a path to the output of the primary target. This way, the publisher
+implementation can remain generic, the `_sphinx_site` rule can just produce
+HTML, and this small script is all that's necessary to combine the two
+together.
+
+
+In `rules_k8s`, this is indeed what `.apply` does:
+[`expand_template`](https://github.com/bazelbuild/rules_k8s/blob/f10e7025df7651f47a76abf1db5ade1ffeb0c6ac/k8s/object.bzl#L213-L241)
+writes a very simple Bash script, based on
+[`apply.sh.tpl`](https://github.com/bazelbuild/rules_k8s/blob/f10e7025df7651f47a76abf1db5ade1ffeb0c6ac/k8s/apply.sh.tpl),
+which runs `kubectl` with the output of the primary target. This script can
+then be build and run with `bazel run :staging.apply`, effectively providing a
+`k8s-apply` command for `k8s_object` targets.
```

Comment:
Ideally, we would have an example on https://github.com/bazelbuild/examples/ and link it from here.

Reply:
I agree that would be nice to have! Do you consider that a blocker for merging this documentation change, since it'll have to be two PRs anyway?

Reply:
Doesn't have to be a blocker. We can always update documentation later.